src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBulkRecordReader.java

```
@@ -143,50 +170,57 @@ public void close() throws IOException {
   }
 
   @VisibleForTesting
-  void setupParser(InputStream queryResponseStream) throws IOException {
-    CSVFormat csvFormat = CSVFormat.DEFAULT.
-      withHeader().
-      withQuoteMode(QuoteMode.ALL).
-      withAllowMissingColumnNames(false);
-
-    csvParser = CSVParser.parse(queryResponseStream, StandardCharsets.UTF_8, csvFormat);
-
-    if (csvParser.getHeaderMap().isEmpty()) {
-      throw new IllegalStateException("Empty response was received from Salesforce, but csv header was expected.");
+  void setupParser() throws IOException, AsyncApiException {
+    if (resultIdIndex < resultIds.length) {
```

This should never be called if there are no more results, so it's better to throw an exception in that case instead of have it be a no-op. Also more readable to check the inverse to remove some nesting and a large if block:

```
if (resultIdIndex >= resultIds.length) {
  throw new IllegalArgumentException(...);
}
...
```

---

fixed